src/tensors/tensor.jl

```diff
@@ -70,89 +67,215 @@ storagetype(::Type{<:TensorMap{T, S, N₁, N₂, A}}) where {T, S, N₁, N₂, A
-    TensorMap{T,S,N₁,N₂,A}(undef, domain → domain)
 
-Construct a `TensorMap` with uninitialized data.
+Construct a `TensorMap` with uninitialized data with elements of type `T`.
 """
-function TensorMap{T}(::UndefInitializer, V::TensorMapSpace{S, N₁, N₂}) where {T, S, N₁, N₂}
-    return TensorMap{T, S, N₁, N₂, Vector{T}}(undef, V)
-end
-function TensorMap{T}(
-        ::UndefInitializer, codomain::TensorSpace{S}, domain::TensorSpace{S}
-    ) where {T, S}
-    return TensorMap{T}(undef, codomain ← domain)
-end
-function Tensor{T}(::UndefInitializer, V::TensorSpace{S}) where {T, S}
-    return TensorMap{T}(undef, V ← one(V))
+TensorMap{T}(::UndefInitializer, V::TensorMapSpace) where {T} =
+    tensormaptype(spacetype(V), numout(V), numin(V), T)(undef, V)
+TensorMap{T}(::UndefInitializer, codomain::TensorSpace, domain::TensorSpace) where {T} =
+    TensorMap{T}(undef, codomain ← domain)
+Tensor{T}(::UndefInitializer, V::TensorSpace) where {T} = TensorMap{T}(undef, V ← one(V))
+
+"""
+    TensorMapWithStorage{T, A}(undef, codomain, domain) where {T, A}
+    TensorMapWithStorage{T, A}(undef, codomain ← domain) where {T, A}
+    TensorMapWithStorage{T, A}(undef, domain → codomain) where {T, A}
+
+Construct a `TensorMap` with uninitialized data stored as `A <: DenseVector{T}`.
+"""
+TensorMapWithStorage{T, A}(::UndefInitializer, V::TensorMapSpace) where {T, A} =
+    tensormaptype(spacetype(V), numout(V), numin(V), A)(undef, V)
+TensorMapWithStorage{T, A}(::UndefInitializer, codomain::TensorSpace, domain::TensorSpace) where {T, A} =
+    TensorMapWithStorage{T, A}(undef, codomain ← domain)
+TensorWithStorage{T, A}(::UndefInitializer, V::TensorSpace) where {T, A} = TensorMapWithStorage{T, A}(undef, V ← one(V))
+
+# raw data constructors
+# ---------------------
+# - dispatch starts with TensorMap{T}(::DenseVector{T}, ...)
+# - select A and map to (TensorMap{T, S, N₁, N₂, A} where {S, N₁, N₂})(::DenseVector{T}, ...)
+# - select S, N₁, N₂ and map to TensorMap{T, S, N₁, N₂, A}(::DenseVector{T}, ...)
+
+"""
+    TensorMap{T}(data::DenseVector{T}, codomain::ProductSpace{S, N₁}, domain::ProductSpace{S, N₂}) where {T, S, N₁, N₂}
+    TensorMap{T}(data::DenseVector{T}, codomain ← domain)
+    TensorMap{T}(data::DenseVector{T}, domain → codomain)
+
+Construct a `TensorMap` from the given raw data.
+This constructor takes ownership of the provided vector, and will not make an independent copy.
+"""
+TensorMap{T}(data::DenseVector{T}, V::TensorMapSpace) where {T} =
+    tensormaptype(spacetype(V), numout(V), numin(V), typeof(data))(data, V)
+TensorMap{T}(data::DenseVector{T}, codomain::TensorSpace, domain::TensorSpace) where {T} =
+    TensorMap{T}(data, codomain ← domain)
+
+"""
+    TensorMapWithStorage{T, A}(data::DenseVector{T}, codomain, domain) where {T, A}
+    TensorMapWithStorage{T, A}(data::DenseVector{T}, codomain ← domain) where {T, A}
+    TensorMapWithStorage{T, A}(data::DenseVector{T}, domain → codomain) where {T, A}
```

Comment:
Even if the `A<:DenseVector{T}` restriction is not necessary in the function definition because it is enforced by the constraints in the type definition, it would be useful to have it here:
```suggestion
    TensorMapWithStorage{T, A}(data::A, codomain, domain) where {T, A<:DenseVector{T}}
    TensorMapWithStorage{T, A}(data::A, codomain ← domain) where {T, A<:DenseVector{T}}
    TensorMapWithStorage{T, A}(data::A, domain → codomain) where {T, A<:DenseVector{T}}
```